repository/src/main/java/org/alfresco/repo/domain/schema/script/DeleteNotExistsV3Executor.java

```diff
@@ -0,0 +1,437 @@
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Properties;
+import java.util.Set;
+
+import javax.sql.DataSource;
+
+import org.alfresco.repo.domain.dialect.Dialect;
+import org.alfresco.repo.domain.dialect.MySQLInnoDBDialect;
+import org.alfresco.util.Pair;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+/**
+ * Same logic as DeleteNotExistsExecuter with the following changes:
+ * <p/>
+ * - filters the queries by unique values
+ * <p/>
+ * - eager close of result sets
+ * <p/>
+ * - we store all the ids in memory and process them from there - the secondary ids are stored in a unique list without
+ * duplicate values.
+ * <p/>
+ * - we only cross 2 sets (the potencial ids to delete from the primary table with the set of all secondary ids in that
+ * range) removing all elements from the second set from the first set
+ * <p/>
+ * - every {pauseAndRecoverBatchSize} rows deleted we close all prepared statements and close the connection and sleep
+ * for {pauseAndRecoverTime} milliseconds. This is necessary to allow the DBMS to perform the background tasks without
+ * load from ACS. When we do not do this and if we are performing millions of deletes, the connection eventually gets
+ * aborted.
+ * 
+ * @author Eva Vasques
+ */
+public class DeleteNotExistsV3Executor extends DeleteNotExistsExecutor
+{
+    private static Log logger = LogFactory.getLog(DeleteNotExistsV3Executor.class);
+
+    public static final String PROPERTY_PAUSE_AND_RECOVER_BATCHSIZE = "system.delete_not_exists.pauseAndRecoverBatchSize";
+    public static final String PROPERTY_PAUSE_AND_RECOVER_TIME = "system.delete_not_exists.pauseAndRecoverTime";
+
+    private Dialect dialect;
+    private final DataSource dataSource;
+    private long pauseAndRecoverTime;
+    private long pauseAndRecoverBatchSize;
+    private boolean pauseAndRecover = false;
+    private int processedCounter = 0;
+
+    public DeleteNotExistsV3Executor(Dialect dialect, Connection connection, String sql, int line, File scriptFile,
+            Properties globalProperties, DataSource dataSource)
+    {
+        super(connection, sql, line, scriptFile, globalProperties);
+        this.dialect = dialect;
+        this.dataSource = dataSource;
+    }
+
+    @Override
+    public void execute() throws Exception
+    {
+        checkProperties();
+
+        String pauseAndRecoverBatchSizeString = globalProperties.getProperty(PROPERTY_PAUSE_AND_RECOVER_BATCHSIZE);
+        pauseAndRecoverBatchSize = pauseAndRecoverBatchSizeString == null ? 500000
+                : Long.parseLong(pauseAndRecoverBatchSizeString);
+
+        String pauseAndRecoverTimeString = globalProperties.getProperty(PROPERTY_PAUSE_AND_RECOVER_TIME);
+        pauseAndRecoverTime = pauseAndRecoverTimeString == null ? 300000 : Long.parseLong(pauseAndRecoverTimeString);
```

Comment:
Similarly 300000 could be a constant - something like `DEFAULT_RECOVERY_PAUSE_MS`.